build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/JarApiComparisonTask.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal;
+
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.TaskAction;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarFile;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+
+/**
+ * This implementation of a jar API comparison uses the "javap" tool to compare
+ * the "signatures" of two different jars. We assume that calling out to javap
```

Comment:
It's over 100ms on my machine to scan a single class. This implementation forks `javap` for every class in the jar. That's not cheap. I think we might want to consider the ASM or other direct bytecode inspection approach sooner rather than later. There are other limitations here that I mention later.

Reply:
If the javap launcher is slow, we can invoke the javap class directly, it's all java code anyway, available in the tools.jar.

Reply:
Let's do some timings and if the overhead is too large we can use `com.sun.tools.javap.JavapTask`.